src/bit-systems/waypoint.ts

```diff
@@ -0,0 +1,209 @@
+
+    const waypoint = spawnPoints[Math.floor(Math.random() * spawnPoints.length)];
+    if (yield* tryOccupyAndSpawn(world, characterController, waypoint)) return true;
+  }
+
+  return false;
+}
+
+function* moveToSpawnPointJob(world: HubsWorld, characterController: CharacterControllerSystem) {
+  if (yield* trySpawnIntoOccupiable(world, characterController)) return;
+
+  const spawnPoints = nonOccupiableSpawnPoints(world);
+  if (spawnPoints.length) {
+    const waypoint = spawnPoints[Math.floor(Math.random() * spawnPoints.length)];
+    moveToWaypoint(world, waypoint, characterController, true);
+  } else {
+    console.warn("Could not find any available spawn points, spawning at the origin.");
+    characterController.enqueueWaypointTravelTo(new Matrix4().identity(), true, {
+      willDisableMotion: false,
+      willDisableTeleporting: false,
+      snapToNavMesh: true,
+      willMaintainInitialOrientation: false
+    });
+  }
+}
+
+let spawnJob: Coroutine | null = null;
+export function moveToSpawnPoint(world: HubsWorld, characterController: CharacterControllerSystem) {
+  spawnJob = coroutine(moveToSpawnPointJob(world, characterController));
+}
+
+function moveToWaypoint(
+  world: HubsWorld,
+  eid: number,
+  characterController: CharacterControllerSystem,
+  instant: boolean
+) {
+  const obj = world.eid2obj.get(eid)!;
+  obj.updateMatrices();
+
+  characterController.enqueueWaypointTravelTo(
+    obj.matrixWorld,
+    instant || !window.APP.store.state.preferences.animateWaypointTransitions, // TODO: Use store-instance
+    {
+      willDisableMotion: !!(Waypoint.flags[eid] & WaypointFlags.willDisableMotion),
+      willDisableTeleporting: !!(Waypoint.flags[eid] & WaypointFlags.willDisableTeleporting),
+      snapToNavMesh: !!(Waypoint.flags[eid] & WaypointFlags.snapToNavMesh),
+      willMaintainInitialOrientation: !!(Waypoint.flags[eid] & WaypointFlags.willMaintainInitialOrientation)
+    }
+  );
+}
+
+const hoveredLeftWaypointQuery = defineQuery([Waypoint, HoveredRemoteLeft]);
+const hoveredRightWaypointQuery = defineQuery([Waypoint, HoveredRemoteRight]);
+
+const exitedOwnedQuery = exitQuery(defineQuery([Owned]));
+
+let preview: Object3D | null;
+export function waypointSystem(
+  world: HubsWorld,
+  characterController: CharacterControllerSystem,
+  sceneIsFrozen: boolean
+) {
+  if (exitedOwnedQuery(world).includes(myOccupiedWaypoint)) {
+    myOccupiedWaypoint = 0;
+  }
+
+  waypointQuery(world).forEach(eid => {
+    if (hasComponent(world, NetworkedWaypoint, eid) && hasComponent(world, Owned, eid) && eid !== myOccupiedWaypoint) {
+      // Inherited this waypoint, clear its occupied state
+      if (NetworkedWaypoint.occupied[eid]) {
+        NetworkedWaypoint.occupied[eid] = 0;
+      }
+    }
+
+    if (hasComponent(world, Interacted, eid)) {
+      if (hasComponent(world, NetworkedWaypoint, eid)) {
+        if (NetworkedWaypoint.occupied[eid]) {
+          console.warn("Can't travel to occupied waypoint...");
```

Comment:
They should ideally not even be clickable in the first place.

Reply:
True, something would have had to go wrong to get to this point.